### DISCUSSION TITLE
Best practice: prop-drilling vs formContext vs slimmed down context

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7553] [Q&A] [answered]
I'm building a proof of concept before migrating from Final-Form. It would be great with some advice.

My custom components will need some props so they can register themselves. At the very least they need `control`. In my spike have prop-drilled so control is available in all of my components. This is not a good solution since it's a long way down the tree to my components. I will also need to marshall some additional functions and data to my components.

The documentation is very helpful and unopinionated in this regard. I coulld need some opinion this time =)

Solutions from the docs:
- useFormContext
- "Smart Form Component"
- prop drilling

I don't want to prop drill.
Smart Form Component is too messy for my taste
useFormContext seems to provide everything, i.e. also leads to rerenders all the time.

Since I only need o send `control` and some of my own data down the tree, I thought I could do this with a cutom, more lightweight context. This context would not hold any frequently updating data, so rerenders would be minimal.

How do others solve this?
Have I missed any solution?

**Comment by Moshyfawn** [comment, marked as answer]
My final nope to "Smart Form Component" solution was when I read [MUI React components library docs on Composition](https://mui.com/guides/composition/#main-content). And while it's great for the flexible library MUI is, working with forms this way is just painful, IMHO. Additionally, I'm using hooks; ascending from `<Form />` components to headless logic functions. It's much better to reuse logic like that.

Prop drilling is a nightmare better left in the past. Tho, if it's gonna be a "once-built-never-touched" system, you can do the drilling. Not saying it's an any good approach, just somewhat valid in that case.

Context was designed to solve the very problem you're facing with deeply nested components that ideally would be doing their own thing, relying on a source of methods for a form. There's [a FormProvider Perfomance section](https://react-hook-form.com/advanced-usage/#FormProviderPerformance) on optimizing re-renders via `React.memo`. Other then that, there's not much to Context. Feels like the right way.

As an option, there's [Recoil](https://recoiljs.org/) and Recoil successors? like [Jotai](https://jotai.org) which allow to basically lift the state "above" the React tree and manage it by subscribing individual components to it. It introduces it's own level of complexity, but it's a nice option too. Tho, I'm not sure it's the way for a deeply nested form inputs / controls.

**Reply by msageryd** [reply]
Thanks again @Moshyfawn 
It really helps to get opinions from others. This time the opinion aligned with my own, which is great.

The thing that trew me off a little from my path was the part in the docs about Smart Form Components. I think it's quite complex, but the fact that it's showcased in the docs made me consider it. I think most devs would want a more opinionated aproach in the docs for guidance. You can do many cool things with Javascript, but that doesn't mean that the code or maintenance of the code is that cool in the long run. (just my 2cents)

**Reply by Moshyfawn** [reply]
I think, having the DX in mind, RHF is trying to put up docs that would have a wiggle room for how to implement a particular feature, since DX is very subjective. That said, the API is more opinionated, which is good. But I see your point. Working on a product, there're times I want to just open docs, see, how it's done in the example / by someone else and just use the solution.

**Reply by bluebill1049** [reply]
as much as I want to put a "best" solution, there is no silver bullet in form application development. It's all case by case.
